### PR TITLE
feat(hierarchical-grouping): Ability to set level of grouping

### PR DESCRIPTION
**Pull request by untitaker**
This PR rebased on top of #26420 allows one to set the current grouping level to a higher one.

~The inverse (decreasing the level, therefore merging issues together) is not implemented in this PR and will probably need a third unmerge message.~ sike it wasn't that hard

**requires** https://github.com/getsentry/snuba/pull/1876 to be merged and deployed

### CODE REVIEW

**Comment by untitaker** [comment]
This PR now also implements reducing the level as it turned out to be less effort than anticipated.

File: src/sentry/api/endpoints/grouping_level_new_issues.py [hunk 1]
```diff
@@ -160,7 +125,7 @@ def _query_snuba(group: Group, unparsed_id: str, offset=None, limit=None):
         # that filtering by timerange (=primary key) is only a little bit
         # faster.
         now = datetime.datetime.now()
-        new_materialized_hash = _get_hash_for_parent_level(group, id, levels_overview)
+        new_materialized_hash = levels_overview.parent_hashes[id]
```
Comment (untitaker) [inline src/sentry/api/endpoints/grouping_level_new_issues.py:128]: This is an unrelated change: I realized that the levels overview now contains the same data that we're computing and caching here.

File: src/sentry/api/endpoints/grouping_levels.py [hunk 2]
```diff
@@ -94,9 +89,13 @@ def get_levels_overview(group):
             [
                 Column("primary_hash"),
                 Function(
-                    "max", [Function("length", [Column("hierarchical_hashes")])], "num_levels"
+                    "argMax",
+                    [
+                        Column("hierarchical_hashes"),
+                        Function("length", [Column("hierarchical_hashes")]),
```
Comment (untitaker) [inline src/sentry/api/endpoints/grouping_levels.py:95]: Instead of sending the postgres hashes to Snuba using `_current_level_expr`, we can simply return the longest array here and do the computation on the clientside.

This query takes up to 40ms in prod. I wonder if there's a faster way to compute what we're computing here, and also how to cache this with proper cache invalidation

File: src/sentry/api/endpoints/grouping_levels.py [hunk 3]
```diff
@@ -115,18 +114,42 @@ def get_levels_overview(group):
 
     fields = res["data"][0]
 
+    if not fields["longest_hierarchical_hashes"]:
```
Comment (untitaker) [inline src/sentry/api/endpoints/grouping_levels.py:117]: empty array => regular grouping applies and we cannot show any sort of levels information or anything based on it

File: src/sentry/unmerge.py [hunk 4]
```diff
@@ -89,25 +98,117 @@ def get_unmerge_key(
...
+        except IndexError:
+            return None
+
+        try:
+            return hierarchical_hashes[self.new_level]
+        except IndexError:
+            return hierarchical_hashes[-1]
+
+    @property
+    def primary_hashes_to_lock(self) -> Collection[str]:
```
Comment (untitaker) [inline src/sentry/unmerge.py:167]: Not sure what we would lock here. We cannot lock primary hash itself as in the case of hierarchical grouping, there may be concurrent unmerges on different groups that contain the same primary_hash (this is not the case for regular unmerge)

For now just allow races, we can figure out a better synchronization mechanism later

File: src/sentry/api/endpoints/grouping_level_details.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    #
+    # We're upserting the hash here to make sure it exists. We have
+    # observed that the materialized hash in postgres is mysteriously lost,
+    # presumably because of secondary grouping or merge/unmerge.
+    grouphash, _created = GroupHash.objects.get_or_create(
+        project_id=group.project_id, hash=levels_overview.current_hash
+    )
+    grouphash.state = GroupHash.State.SPLIT
+    grouphash.group_id = None
+    grouphash.save()
```
Comment (jjbayer) [inline src/sentry/api/endpoints/grouping_level_details.py:81]: Is there any threat of a race condition between `get` and `save`?

File: src/sentry/api/endpoints/grouping_levels.py [hunk 6]
```diff
@@ -115,18 +114,42 @@ def get_levels_overview(group):
...
+        for hash, in GroupHash.objects.filter(project=group.project, group=group).values_list(
+            "hash"
+        )
+    }
+
+    current_level = max(
+        i
+        for i, hash in enumerate(fields["longest_hierarchical_hashes"])
+        if hash in materialized_hashes
+    )
```
Comment (jjbayer) [inline src/sentry/api/endpoints/grouping_levels.py:131]: Is there any danger of `materialized_hashes` being empty and `max` failing here?

File: src/sentry/api/endpoints/grouping_level_details.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+                    "argMax",
+                    [
+                        Function(
+                            "arraySlice",
+                            [
+                                Column("hierarchical_hashes"),
+                                id + 1,
+                            ],
+                        ),
+                        Column("timestamp"),
```
Comment (jjbayer) [inline src/sentry/api/endpoints/grouping_level_details.py:141]: Could it happen that `hierarchical_hashes` is shorter for some events (i.e. the call stack is less deep) than for others, and we then miss some hashes that should be reset?

File: src/sentry/api/endpoints/grouping_level_details.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+    #
+    # If we move this into a paginating celery task we need to consider that
+    # we're iterating over a table that is also modified by the unmerge task at
+    # the same time.
+    for row in snuba.raw_snql_query(
+        query, referrer="api.grouping_levels_details.decrease_level.get_group_ids"
+    )["data"]:
+        group_id = row["group_id"]
+
+        replacement = HierarchicalUnmergeReplacement(
```
Comment (jjbayer) [inline src/sentry/api/endpoints/grouping_level_details.py:175]: It confused me for a moment that the class is called `UnmergeReplacement` regardless of whether the operation is increase or decrease.

**Review by jjbayer — APPROVED** [review]

File: src/sentry/api/endpoints/grouping_level_details.py (hunk 5 above)
Comment (untitaker) [reply src/sentry/api/endpoints/grouping_level_details.py:103]: we can't really fix this race, without a row there is no row lock we could acquire. changed to update_or_create

File: src/sentry/api/endpoints/grouping_level_details.py (hunk 7 above)
Comment (untitaker) [reply src/sentry/api/endpoints/grouping_level_details.py]: good point, I rewrote this piece of code such that all hashes are returned here for sure

File: src/sentry/api/endpoints/grouping_levels.py (hunk 6 above)
Comment (untitaker) [reply src/sentry/api/endpoints/grouping_levels.py:131]: yes, happened multiple times to me in development, but it should be a crash. it's invalid internal state

File: src/sentry/api/endpoints/grouping_level_details.py (hunk 8 above)
Comment (untitaker) [reply src/sentry/api/endpoints/grouping_level_details.py]: This is now gone after a discussion on the Snuba PR

File: src/sentry/api/endpoints/grouping_level_details.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+    # GroupingLevelNewIssuesEndpoint.
+
+    query = (
+        Query("events", Entity("events"))
+        .set_select(
+            [
+                Function(
+                    "arrayDistinct", [Function("groupArray", [Column("group_id")])], "group_ids"
+                ),
+                Function(
```
Comment (untitaker) [inline src/sentry/api/endpoints/grouping_level_details.py:141]: @tkaemming this query is nuts but I believe it may be the fastest way to compute two unrelated queries at the same time: `SELECT distinct group_id ...` and `SELECT distinct arrayJoin(hierarchical_hashes) ...`. I suppose it would be better to send off two queries in a threadpool?
